eva/optimizer/optimizer_tasks.py

```diff
@@ -85,172 +74,188 @@ def root_expr(self, expr: GroupExpression):
+                 optimizer_context: OptimizerContext,
+                 explore: bool):
+        self.root_expr = root_expr
+        self.explore = explore
+        super().__init__(optimizer_context,
                          OptimizerTaskType.OPTIMIZE_EXPRESSION)
 
     def execute(self):
-        implementation_rules = RulesManager().implementation_rules
+        all_rules = RulesManager().logical_rules
+        # if exploring, we don't need to consider implementation rules
+        if not self.explore:
+            all_rules.extend(RulesManager().implementation_rules)
+
         valid_rules = []
-        for rule in implementation_rules:
+        for rule in all_rules:
             if rule.top_match(self.root_expr.opr):
                 valid_rules.append(rule)
 
-        sorted(valid_rules, key=lambda x: x.promise(), reverse=True)
+        sorted(valid_rules, key=lambda x: x.promise())
+
         for rule in valid_rules:
-            binder = Binder(self.root_expr, rule.pattern,
-                            self.optimizer_context.memo)
-            for match in iter(binder):
-                if not rule.check(match, self.optimizer_context):
-                    continue
-                LoggingManager().log('In Optimize physical expression,'
-                                     'Rule {} matched for {}'
-                                     .format(rule, self.root_expr),
-                                     LoggingLevel.INFO)
-                after = rule.apply(match, self.optimizer_context)
-                new_expr = GroupExpression(
-                    after, self.root_expr.group_id, self.root_expr.children)
-                # LoggingManager().log('After rewiting {}'.format(new_expr),
-                #                     LoggingLevel.INFO)
-                self.optimizer_context.memo.add_group_expr(new_expr)
-                # Optimize inputs for this physical expr
-                self.optimizer_context.task_stack.push(
-                    OptimizeInputs(new_expr, self.optimizer_context)
-                )
+            # apply the rule
+            self.optimizer_context.task_stack.push(
+                ApplyRule(rule, self.root_expr,
+                          self.optimizer_context, self.explore)
+            )
 
-            # Optimize the child groups
-            for child_id in self.root_expr.children:
-                self.optimizer_context.task_stack.push(
-                    OptimizeGroup(child_id, self.optimizer_context)
-                )
+            # explore the input group if necessary
+            for idx, child in enumerate(rule.pattern.children):
+                if len(child.children):
+                    child_grp_id = self.root_expr.children[idx]
+                    group = self.optimizer_context.memo.get_group_by_id(
+                        child_grp_id)
+                    self.optimizer_context.task_stack.push(
+                        ExploreGroup(group, self.optimizer_context)
+                    )
```

Comment:
Need some help to understand the ExploreGroup.

Reply:
This is to ensure that the child groups are explored. For example, cases where you add a new logical equivalent expression.